### PR TITLE
feat: implement path transforms

### PR DESCRIPTION
**Pull request by thesayyn**
Adds support for path transforming for the tar rule. Effectively making pkg_tar do everything pkg_tar can do and more. 

### Type of change

- New feature or functionality (change which adds functionality)

**For changes visible to end-users**
- Relevant documentation has been updated

### Test plan

- New test cases added

### CODE REVIEW

File: lib/tests/tar/BUILD.bazel [hunk 1]
```diff
@@ -199,3 +199,30 @@ diff_test(
...
+        "pkg/debian-binary$": "debian-binary",
+    },
+)
+
+assert_tar_listing(
+    name = "test_transform",
+    actual = "transform",
+    expected = [
+        #
+        # TODO: https://github.com/aspect-build/bazel-lib/issues/625
```
Comment (jbedard) [inline lib/tests/tar/BUILD.bazel:224]: What should it look like?

File: lib/tests/tar/BUILD.bazel [hunk 2]
```diff
@@ -199,3 +199,30 @@ diff_test(
...
 )
+
+tar(
+    name = "transform",
+    srcs = [
+        "pkg/debian-binary",
+        "src_file",
+    ],
+    out = "transform.tar",
+    transform = {
```
Comment (jbedard) [inline lib/tests/tar/BUILD.bazel:210]: This doesn't cover the case having no `^` or `$`?

File: lib/tar.bzl [hunk 3]
```diff
@@ -23,7 +23,27 @@ load("@bazel_skylib//rules:write_file.bzl", "write_file")
 load("//lib/private:tar.bzl", "tar_lib", _tar = "tar")
 
 mtree_spec = rule(
-    doc = "Create an mtree specification to map a directory hierarchy. See https://man.freebsd.org/cgi/man.cgi?mtree(8)",
+    doc = """Create an mtree specification to map a directory hierarchy. See https://man.freebsd.org/cgi/man.cgi?mtree(8)
+
+Supports `$` and `^` RegExp tokens, which may be used together.
```
Comment (jbedard) [inline lib/tar.bzl:28]: Is this `$`+`^` a standard from somewhere? Having the starts/ends-with but no other regex/glob?

File: lib/tests/tar/BUILD.bazel (hunk 2 above)
Comment (thesayyn) [reply lib/tests/tar/BUILD.bazel:210]: right. will add that.

File: lib/tests/tar/BUILD.bazel (hunk 1 above)
Comment (thesayyn) [reply lib/tests/tar/BUILD.bazel:224]:
```
"drwxr-xr-x  0 0      0           0 Jan  1  2023 package",
"-rwxr-xr-x  0 0      0           0 Jan  1  2023 package/debian-binary",
"-rwxr-xr-x  0 0      0          21 Jan  1  2023 package/PKGINFO",
```

File: lib/tar.bzl (hunk 3 above)
Comment (thesayyn) [reply lib/tar.bzl:28]: It's standard regex tokens. https://regex101.com/r/iF7PKY/1

Comment (thesayyn) [reply lib/tar.bzl:28]: oh, i can only support those tokens as they are trivial and easy to implement.

Comment (thesayyn) [reply lib/tar.bzl:28]: and since we are applying these transforms serially, anything is possible.

File: lib/tests/tar/BUILD.bazel (hunk 2 above)
Comment (thesayyn) [reply lib/tests/tar/BUILD.bazel:210]: done

File: lib/tar.bzl (hunk 3 above)
Comment (jbedard) [reply lib/tar.bzl:28]: I mean is the idea of only supporting those copied from somewhere else. Are we trying to support the same thing as a rule we're replacing? Or is this new functionality now only in `bazel-lib`?

Comment (jbedard) [reply lib/tar.bzl:28]: If this isn't copying from somewhere else... why regex instead of glob like everywhere else in bazel-lib?

Comment (thesayyn) [reply lib/tar.bzl:28]: This is only available in bazel-lib. This is for transforming paths, whereas globbing is used for matching paths. do we do something similar with globs here?

File: lib/private/tar.bzl [hunk 4]
```diff
@@ -139,12 +136,32 @@ def _mtree_line(file, content, type, uid = "0", gid = "0", time = "1672560000",
...
+            if match.removeprefix("^").removesuffix("$") == path:
+                path = replace
+        elif match.startswith("^"):
+            if path.startswith(match.removeprefix("^")):
+                path = "".join([replace, path.removeprefix(match.removeprefix("^"))])
+        elif match.endswith("$"):
+            if path.endswith(match.removesuffix("$")):
+                path = "".join([path.removesuffix(match.removesuffix("$")), replace])
+        else:
+            path = path.replace(match, replace)
```
Comment (jbedard) [inline lib/private/tar.bzl:152]: Does `replace` replace all?

File: lib/tar.bzl (hunk 3 above)
Comment (jbedard) [reply lib/tar.bzl:28]: After discussing it we figured globs is not the right solution, but I'd like this public API to be reviewed by @alexeagle and maybe others before we introduce an API like this.

File: lib/private/tar.bzl (hunk 4 above)
Comment (thesayyn) [reply lib/private/tar.bzl:152]: yes

Comment (jbedard) [reply lib/private/tar.bzl:152]: Do we have a test for that?

File: lib/tests/tar/BUILD.bazel (hunk 1 above)
Comment (jbedard) [reply lib/tests/tar/BUILD.bazel:224]: Should you explain that in the comment? Even should "should also contain X, see: URL"

File: lib/private/tar.bzl (hunk 4 above)
Comment (thesayyn) [reply lib/private/tar.bzl:152]: hmmm, no. i am gonna add a unit test for it now

File: lib/tests/tar/BUILD.bazel (hunk 2 above)
Comment (alexeagle) [inline lib/tests/tar/BUILD.bazel:210]: I think we should spend a couple more weeks to be very thoughtful in what public API we introduce here

**Comment by alexeagle** [comment]
I think we determined that https://github.com/vbatts/go-mtree/ is the utility we want to bring for users to do this.

**Pull request closed without merging**